sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobBatch.java

```diff
@@ -0,0 +1,413 @@
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+/**
+ * This class provides
+ */
+public final class BlobBatch {
+    private static final String BATCH_REQUEST_CONTENT_ID = "Batch-Request-Content-Id";
+    private static final String CONTENT_ID = "Content-Id";
+    private static final String BATCH_BOUNDARY_TEMPLATE = "batch_%s";
+    private static final String CONTENT_TYPE = "Content-Type: application/http";
+    private static final String CONTENT_TRANSFER_ENCODING = "Content-Transfer-Encoding: binary";
+    private static final String CONTENT_ID_TEMPLATE = "Content-ID: %d";
+    private static final String HTTP_VERSION = "HTTP/1.1";
+    private static final String OPERATION_TEMPLATE = "%s %s %s";
+    private static final String HEADER_TEMPLATE = "%s: %s";
+
+    private static final int[] EXPECTED_DELETE_STATUS_CODES = { 202 };
+    private static final int[] EXPECTED_SET_TIER_STATUS_CODES = { 200, 202 };
+
+    private final ClientLogger logger = new ClientLogger(BlobBatch.class);
+
+    private final String accountUrl;
+    private final HttpPipeline batchPipeline;
+
+    private final Deque<Mono<? extends Response>> batchOperationQueue;
+    private final List<ByteBuffer> batchRequest;
+    private final Map<Integer, BlobBatchOperationResponse> batchMapping;
+
+    private final AtomicInteger contentId;
+    private final String batchBoundary;
+
+    private BlobBatchType batchType;
+
+    public BlobBatch(BlobServiceClient client) {
+        this(client.getAccountUrl(), client.getHttpPipeline());
+    }
+
+    public BlobBatch(BlobServiceAsyncClient client) {
+        this(client.getAccountUrl(), client.getHttpPipeline());
+    }
+
+    BlobBatch(String accountUrl, HttpPipeline pipeline) {
+        this.contentId = new AtomicInteger(0);
+        this.batchBoundary = String.format(BATCH_BOUNDARY_TEMPLATE, UUID.randomUUID());
+
+        this.accountUrl = accountUrl;
+
+        boolean batchHeadersPolicySet = false;
+        List<HttpPipelinePolicy> policies = new ArrayList<>();
+        for (int i = 0; i < pipeline.getPolicyCount(); i++) {
+            HttpPipelinePolicy policy = pipeline.getPolicy(i);
+
+            // Don't want these policies in the dummy pipeline.
+            if (policy instanceof HttpLoggingPolicy
+                || policy instanceof RetryPolicy
+                || policy instanceof RequestRetryPolicy
+                || policy instanceof UserAgentPolicy) {
+                continue;
+            }
+
+            if (policy instanceof SharedKeyCredentialPolicy) {
+                batchHeadersPolicySet = true;
+                // The batch policy needs to be added before the SharedKey policy to run preparation cleanup.
+                policies.add(new BatchHeadersPolicy());
+            }
+
+            policies.add(pipeline.getPolicy(i));
+        }
```

Comment:
This is slightly traumatising.

Reply:
I think this could be simplified to just adding the batch policy before the credential policy.

Reply:
As long as it still gets the signing from the credentials but doesn't get retries

Reply:
Remove the removal of policies, retry policies shouldn't trigger here given the response is `Mono.empty()`